src/thor/expansion_action.cc

```diff
@@ -78,43 +118,56 @@ std::string thor_worker_t::expansion(Api& request) {
-    const auto* edge = tile->directededge(edgeid);
-    // unfortunately we have to call this before checking if we can skip
-    // else the tile could change underneath us when we get the opposing
-    auto shape = tile->edgeinfo(edge).shape();
-    auto names = tile->edgeinfo(edge).GetNames();
-    auto is_forward = edge->forward();
-
-    // if requested, skip this edge in case its opposite edge has been added
-    // before (i.e. lower cost) else add this edge's id to the lookup container
-    if (skip_opps) {
-      auto opp_edgeid = reader.GetOpposingEdgeId(edgeid, tile);
-      if (opp_edgeid && opp_edges.count(opp_edgeid))
-        return;
-      opp_edges.insert(edgeid);
-    }
-
-    if (!edge->forward())
-      std::reverse(shape.begin(), shape.end());
-    Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
-
-    writeExpansionProgress(expansion, edgeid, prev_edgeid, shape, exp_props, status, duration,
-                           distance, cost);
-  };
+  auto track_expansion =
+      [&expansion, &opp_edges, &gen_factor, &skip_opps, &exp_props, &algo, &edge_state,
+       &dedupe](baldr::GraphReader& reader, baldr::GraphId edgeid, baldr::GraphId prev_edgeid,
+                const char* algorithm = nullptr,
+                const Expansion_EdgeStatus status = Expansion_EdgeStatus_reached,
+                const float duration = 0.f, const uint32_t distance = 0, const float cost = 0.f) {
+        algo = algorithm;
+
+        auto tile = reader.GetGraphTile(edgeid);
+        if (tile == nullptr) {
+          LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                    std::to_string(edgeid.Tile_Base()));
+          return;
+        }
+
+        // if requested, skip this edge in case its opposite edge has been added
+        // before (i.e. lower cost) else add this edge's id to the lookup container
+        if (skip_opps) {
+          auto opp_tile = tile;
+          auto opp_edgeid = reader.GetOpposingEdgeId(edgeid, opp_tile);
+          if (opp_edgeid && opp_edges.count(opp_edgeid))
+            return;
+          opp_edges.insert(edgeid);
+        }
+
+        const auto* edge = tile->directededge(edgeid);
+        auto shape = tile->edgeinfo(edge).shape();
+
+        if (!edge->forward())
+          std::reverse(shape.begin(), shape.end());
+        Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
+        if (dedupe) {
+          uint8_t stage = 1 << status;
+          if (edge_state.contains(edgeid)) {
+            auto edge_stages = edge_state.at(edgeid).stages_mask;
+            // Keep only properties of last/highest status of edge, but update what stages the edge
+            // has seen
+            if (!expansion_properties_t::is_latest_status(edge_stages, stage)) {
+              edge_state.at(edgeid).stages_mask |= stage;
+              return;
+            }
+            stage |= edge_state.at(edgeid).stages_mask;
+          }
+          edge_state[edgeid] =
```

Comment:
i very much appreciate that we dont fill up this large map when not deduping, thank you for that!